{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f828f4ec",
   "metadata": {},
   "source": [
    "# 1. Loading the PETase dataset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e367f141",
   "metadata": {},
   "outputs": [],
   "source": [
    "#we will now load \"masterdb.csv\" found under data\n",
    "import pandas as pd \n",
    "import os \n",
    "path = \"data/masterdb.tsv\"\n",
    "df = pd.read_csv(path,sep=\"\\t\")\n",
    "df.columns = (\n",
    "    df.columns\n",
    "    .str.strip()\n",
    "    .str.lower()\n",
    "    .str.replace(\" \", \"_\")\n",
    "    .str.replace(\"(\", \"\")\n",
    "    .str.replace(\")\", \"\")\n",
    "    .str.replace(\"?\", \"\")\n",
    ")\n",
    "df2 = pd.DataFrame()\n",
    "df2[\"id\"]=df[\"name\"]\n",
    "df2[\"sequence\"] = df[\"protein_sequence\"].astype(str)\n",
    "print(df2)\n",
    "\n",
    "def dftofasta(df,outfile):\n",
    "    with open(outfile,\"w\") as f:\n",
    "        for index,row in df.iterrows():\n",
    "            f.write(f\">{row['id']}\\n\")\n",
    "            f.write(f\"{row['sequence']}\\n\")\n",
    "    return outfile \n",
    "\n",
    "#dftofasta(df2,\"data/masterdb.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5d3b47a",
   "metadata": {},
   "source": [
    "# 2. Loading external stability and expression datasets\n",
    "\n",
    "**NESG Solubility** \n",
    "(https://loschmidt.chemi.muni.cz/soluprot/?page=download)\n",
    "* 10k proteins\n",
    "* Labels: exp, sol, uniprot id or local ID \n",
    "* Units: integer \n",
    "\n",
    "**Soluprot Solubility**\n",
    "(https://loschmidt.chemi.muni.cz/soluprot/?page=download)\n",
    "* 11k training, 3k test\n",
    "* Label: solubility, number IDs with no conversion map (has seq)\n",
    "* Unit: 0/1\n",
    "\n",
    "**Price Solubility**\n",
    "(https://pmc.ncbi.nlm.nih.gov/articles/PMC3372292/)\n",
    "* 7k proteins \n",
    "* Label: usability. uniprot id\n",
    "* Unit: 0/1\n",
    "\n",
    "**PSI Solubility** \n",
    "(https://academic.oup.com/bioinformatics/article/36/18/4691/5860015?login=false)\n",
    "* 11k proteins\n",
    "* Label: solubility, Aa0000 ID scheme (has seq)\n",
    "* Unit: 0/1\n",
    "\n",
    "**Meltome Stability** \n",
    "(https://meltomeatlas.proteomics.wzw.tum.de/master_meltomeatlasapp/)\n",
    "* 1M variants \n",
    "* Label: temperature, meltpoint, fold_change, uniprot id \n",
    "\n",
    "**FireprotDB Stability** \n",
    "(https://loschmidt.chemi.muni.cz/fireprotdb/)\n",
    "* 53k variants\n",
    "* Label: ddG, dTm, pH, Tm, mutation_effect, uniprot id \n",
    "\n",
    "**ThermomutDB Stability**\n",
    "(https://biosig.lab.uq.edu.au/thermomutdb/downloads)\n",
    "* 12k variants\n",
    "* Label: pH, ddG, temperature, dTm, uniprot/pdb id \n",
    "\n",
    "**CAFA** \n",
    "(https://www.kaggle.com/competitions/cafa-5-protein-function-prediction/code)\n",
    "* 142k variants\n",
    "\n",
    "**Novozyme**\n",
    "(https://www.kaggle.com/code/jinyuansun/eda-and-finetune-esm)\n",
    "* 31k variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c9baf06",
   "metadata": {},
   "source": [
    "**Protsol Solubility**\n",
    "(https://huggingface.co/datasets/AI4Protein/ProtSolM)\n",
    "* 71k proteins\n",
    "* Label: solubility, no ID but has sequence\n",
    "* Unit: 0/1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb14bb41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#LOADING ALL BENCHMARK DATASETS AND MERGE IN ONE\n",
    "import pandas as pd\n",
    "import json\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "\n",
    "PSI_PATH      = \"data/temp/sol_benchmark/PSI_Biology_solubility_trainset.csv\"\n",
    "psi_detail_path = \"data/temp/sol_benchmark/PSI_all_data_esol.tab\"\n",
    "#note that 1988/11000 sequences matched to a proteome gene, with moist being lower than 50% similarity, so we drop PSI \n",
    "NESG_PATH     = \"data/temp/sol_benchmark/nesg/nesg.csv\"\n",
    "nesg_fasta_path = \"data/temp/sol_benchmark/nesg/nesg.fasta\"\n",
    "PRICE_PATH    = \"data/temp/sol_benchmark/Price_usability_trainset.csv\"\n",
    "price_fasta=\n",
    "soluprot_train_path = \"data/temp/sol_benchmark/soluprot_data/training_set.csv\"\n",
    "soluprot_test_path = \"data/temp/sol_benchmark/soluprot_data/test_set.csv\" \n",
    "meltome_path = \"data/temp/stab_benchmark/meltome_cross-species.csv\"\n",
    "fireprot_path = \"data/temp/stab_benchmark/fireprotdb_results_stability.csv\"\n",
    "thermomutdb_path = \"data/temp/stab_benchmark/thermomutdb.json\"\n",
    "protsol_train_path = \"data/temp/protsolm_data/protsolm_train.csv\"\n",
    "protsol_test_path = \"data/temp/protsolm_data/protsolm_test.csv\"\n",
    "\n",
    "\n",
    "def read_fasta_dict(path: str):\n",
    "    seqs = {}\n",
    "    with open(path) as fh:\n",
    "        for header, seq in SimpleFastaParser(fh):\n",
    "            sid = header.split()[0].strip()\n",
    "            seqs[sid] = seq.strip()\n",
    "    return seqs\n",
    "\n",
    "def load_nesg(csv_path: str, fasta_path: str) -> pd.DataFrame:\n",
    "    df = pd.read_csv(csv_path)  # uses CSV header row directly: id, exp, sol\n",
    "    seqs = read_fasta_dict(fasta_path)\n",
    "    #\"sid\" \"usability\" \"fasta\" \n",
    "    df[\"sequence\"] = df[\"id\"].map(seqs)\n",
    "    return df \n",
    "\n",
    "def load_psi(csv_path: str,psi_detail_path) -> pd.DataFrame:\n",
    "    df = pd.read_csv(csv_path)\n",
    "    psi_all = pd.read_csv(psi_detail_path, sep=\"\\t\")\n",
    "    df = df.merge(psi_all, on=\"sid\", how=\"left\")\n",
    "    df[\"sequence\"] = df[\"fasta\"]\n",
    "    return df\n",
    "\n",
    "def load_price(csv_path: str) -> pd.DataFrame:\n",
    "    df = pd.read_csv(csv_path)\n",
    "    df[\"sequence\"] = df[\"fasta\"]\n",
    "    # \"sid\" \"usability\" \"fasta\" \n",
    "    return df\n",
    "\n",
    "def load_soluprot(train_csv: str, test_csv: str) -> pd.DataFrame:\n",
    "    test_fasta = read_fasta_dict(\"data/temp/sol_benchmark/soluprot_data/test_set.fasta\")\n",
    "    train_fasta = read_fasta_dict(\"data/temp/sol_benchmark/soluprot_data/training_set.fasta\")\n",
    "    fasta = {**train_fasta, **test_fasta}\n",
    "    df1 = pd.read_csv(train_csv)\n",
    "    df2 = pd.read_csv(test_csv)\n",
    "    df = pd.concat([df1, df2], ignore_index=True)\n",
    "    df[\"sequence\"] = df[\"sid\"].astype(str).map(fasta)    # \"sid\" \"solubility\"\n",
    "    csv_ids = set(df[\"sid\"].astype(str))\n",
    "    fasta_ids = set(fasta.keys())\n",
    "\n",
    "    return df\n",
    "\n",
    "def load_meltome(csv_path: str) -> pd.DataFrame:\n",
    "    df = pd.read_csv(csv_path)\n",
    "    df[\"sequence\"] = None\n",
    "    # \"Protein_ID\" \"gene_name\" \"fold_change\" \"meltPoint\" \"temperature\" \"sequence\"\n",
    "    return df\n",
    "\n",
    "def load_fireprot(csv_path: str) -> pd.DataFrame:\n",
    "    df = pd.read_csv(csv_path)\n",
    "    # \"uniprot_id\" \"pdb_id\" \"muutation\" \"ddG\" \"dTm\" \"pH\" \"tm\" \"mutation_effect\" \"sequence\"\n",
    "    return df \n",
    "\n",
    "def load_thermomut(json_path: str) -> pd.DataFrame:\n",
    "    with open(json_path) as fh:\n",
    "        data = json.load(fh)\n",
    "    df = pd.DataFrame(data)\n",
    "    # \"uniprot\" \"ph\" \"ddg\" \"temperature\" \"dtm\" \"PDB_wild\" \"pdb_mutant\" \"mutation_code\"\n",
    "    return df\n",
    "\n",
    "def load_protsolm(train_csv: str, test_csv: str) -> pd.DataFrame:\n",
    "    df1 = pd.read_csv(train_csv)\n",
    "    df2 = pd.read_csv(test_csv)\n",
    "    df = pd.concat([df1, df2], ignore_index=True)\n",
    "    # \"aa_seq\" \"detail\"\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "06426298",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\justi\\AppData\\Local\\Temp\\ipykernel_27684\\2242562371.py:65: DtypeWarning: Columns (1,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(csv_path)\n",
      "C:\\Users\\justi\\AppData\\Local\\Temp\\ipykernel_27684\\2242562371.py:71: DtypeWarning: Columns (23,24,25,26,29) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(csv_path)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              id  exp  sol                                           sequence  \\\n",
      "0        AR3338B  5.0  0.0  MSVHKLTDLRDNSTNWKINVKILSIWNHPPNSHGEITTMILHDDKN...   \n",
      "1        AR3347A  0.0  0.0  IPFDYIVEKTVSTGVLVDVIGALLEVGNLTEDYRGLKLPFKIMDQY...   \n",
      "2        AR3353C  4.0  3.0  MEEERRDDYKFLRIQDAFKALHLHVNLIGVIVELGFSNGSDCSCTL...   \n",
      "3        AR3354C  0.0  0.0  ALLRRFIGQKVRTVIQVTGSEIGSVVGKSTDDLQIVVRGSSPPSPL...   \n",
      "4        AR3358A  5.0  2.0  MAASFAFLRDVRPYKTSWRVQVKVLHSWCQYTNMTGETLKLVLVNS...   \n",
      "...          ...  ...  ...                                                ...   \n",
      "1281497      NaN  NaN  NaN                                                NaN   \n",
      "1281498      NaN  NaN  NaN                                                NaN   \n",
      "1281499      NaN  NaN  NaN                                                NaN   \n",
      "1281500      NaN  NaN  NaN                                                NaN   \n",
      "1281501      NaN  NaN  NaN                                                NaN   \n",
      "\n",
      "         sid  Usability|0=NotUsable|1=Usable fasta  Partition  solubility  \\\n",
      "0        NaN                             NaN   NaN        NaN         NaN   \n",
      "1        NaN                             NaN   NaN        NaN         NaN   \n",
      "2        NaN                             NaN   NaN        NaN         NaN   \n",
      "3        NaN                             NaN   NaN        NaN         NaN   \n",
      "4        NaN                             NaN   NaN        NaN         NaN   \n",
      "...      ...                             ...   ...        ...         ...   \n",
      "1281497  NaN                             NaN   NaN        NaN         NaN   \n",
      "1281498  NaN                             NaN   NaN        NaN         NaN   \n",
      "1281499  NaN                             NaN   NaN        NaN         NaN   \n",
      "1281500  NaN                             NaN   NaN        NaN         NaN   \n",
      "1281501  NaN                             NaN   NaN        NaN         NaN   \n",
      "\n",
      "        run_name  ... dtm pam250  length uniprot  PMID  PIR_ID  \\\n",
      "0            NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "1            NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "2            NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "3            NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "4            NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "...          ...  ...  ..    ...     ...     ...   ...     ...   \n",
      "1281497      NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "1281498      NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "1281499      NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "1281500      NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "1281501      NaN  ... NaN    NaN     NaN     NaN   NaN     NaN   \n",
      "\n",
      "                      name label    detail  \\\n",
      "0                      NaN   NaN       NaN   \n",
      "1                      NaN   NaN       NaN   \n",
      "2                      NaN   NaN       NaN   \n",
      "3                      NaN   NaN       NaN   \n",
      "4                      NaN   NaN       NaN   \n",
      "...                    ...   ...       ...   \n",
      "1281497  test_protein_7839   1.0  SoluProt   \n",
      "1281498  test_protein_7840   1.0  SoluProt   \n",
      "1281499  test_protein_7841   1.0  SoluProt   \n",
      "1281500  test_protein_7842   1.0  SoluProt   \n",
      "1281501  test_protein_7843   1.0  SoluProt   \n",
      "\n",
      "                                                    aa_seq  \n",
      "0                                                      NaN  \n",
      "1                                                      NaN  \n",
      "2                                                      NaN  \n",
      "3                                                      NaN  \n",
      "4                                                      NaN  \n",
      "...                                                    ...  \n",
      "1281497  MPLTTNTILICAGGACISAGEKSVKDAFEEELRKYGLDEVVRVIET...  \n",
      "1281498  MPKNKILIYLLSTTLVLPTLVSPTAYADTPQKDTTAKTTSHDSKKS...  \n",
      "1281499  MSKIVEVSKSPYAKLKTIPVGSVRVNGFMGKYFETLVNVTLPTQYD...  \n",
      "1281500  MTKIFIRDDNGGTEMSQDPFQEREAEKYANPIPSREFILEHLTKRE...  \n",
      "1281501  MVTSSSMRRCRGALSLSFWIFLVLRSVLSRPVFGAAAGPAVQSARQ...  \n",
      "\n",
      "[1281502 rows x 95 columns] (1281502, 95)\n"
     ]
    }
   ],
   "source": [
    "merged = pd.concat([\n",
    "    load_nesg(NESG_PATH, nesg_fasta_path),\n",
    "    load_price(PRICE_PATH),\n",
    "    load_soluprot(soluprot_train_path, soluprot_test_path),\n",
    "    load_meltome(meltome_path),\n",
    "    load_fireprot(fireprot_path),\n",
    "    load_thermomut(thermomutdb_path),\n",
    "    load_protsolm(protsol_train_path, protsol_test_path)\n",
    "], ignore_index=True)\n",
    "\n",
    "print(merged, merged.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cda2ad4",
   "metadata": {},
   "source": [
    "# 3. The activity label of the PETase dataset \n",
    "\n",
    "**Docking** \n",
    "\n",
    "**PET catalysis rate** \n",
    "\n",
    "**PET-specific biophysical features** \n",
    "\n",
    "**Computational chemistry of PET and using it as a feature**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "364efbba",
   "metadata": {},
   "source": [
    "# 4. Fine-tuning esm2 model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "236b8815",
   "metadata": {},
   "source": [
    "# 5. Fine Tuning esm3 model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3bb20a9",
   "metadata": {},
   "source": [
    "# 6. Graph Neural Network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dd63680",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
