{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f828f4ec",
   "metadata": {},
   "source": [
    "# 1. Loading the PETase dataset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e367f141",
   "metadata": {},
   "outputs": [],
   "source": [
    "#we will now load \"masterdb.csv\" found under data\n",
    "import pandas as pd \n",
    "import os \n",
    "path = \"data/masterdb.tsv\"\n",
    "df = pd.read_csv(path,sep=\"\\t\")\n",
    "df.columns = (\n",
    "    df.columns\n",
    "    .str.strip()\n",
    "    .str.lower()\n",
    "    .str.replace(\" \", \"_\")\n",
    "    .str.replace(\"(\", \"\")\n",
    "    .str.replace(\")\", \"\")\n",
    "    .str.replace(\"?\", \"\")\n",
    ")\n",
    "df2 = pd.DataFrame()\n",
    "df2[\"id\"]=df[\"name\"]\n",
    "df2[\"sequence\"] = df[\"protein_sequence\"].astype(str)\n",
    "print(df2)\n",
    "\n",
    "def dftofasta(df,outfile):\n",
    "    with open(outfile,\"w\") as f:\n",
    "        for index,row in df.iterrows():\n",
    "            f.write(f\">{row['id']}\\n\")\n",
    "            f.write(f\"{row['sequence']}\\n\")\n",
    "    return outfile \n",
    "\n",
    "#dftofasta(df2,\"data/masterdb.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5d3b47a",
   "metadata": {},
   "source": [
    "# 2. Loading external stability and expression datasets\n",
    "\n",
    "**NESG Solubility** \n",
    "(https://loschmidt.chemi.muni.cz/soluprot/?page=download)\n",
    "* 10k proteins\n",
    "* Labels: exp, sol, uniprot id \n",
    "* Units: integer \n",
    "\n",
    "**Soluprot Solubility**\n",
    "(https://loschmidt.chemi.muni.cz/soluprot/?page=download)\n",
    "* 11k training, 3k test\n",
    "* Label: solubility, number IDs with no conversion map (has seq)\n",
    "* Unit: 0/1\n",
    "\n",
    "**Price Solubility**\n",
    "(https://pmc.ncbi.nlm.nih.gov/articles/PMC3372292/)\n",
    "* 7k proteins \n",
    "* Label: usability. uniprot id\n",
    "* Unit: 0/1\n",
    "\n",
    "**PSI Solubility** \n",
    "(https://academic.oup.com/bioinformatics/article/36/18/4691/5860015?login=false)\n",
    "* 11k proteins\n",
    "* Label: solubility, Aa0000 ID scheme (has seq)\n",
    "* Unit: 0/1\n",
    "\n",
    "**Meltome Stability** \n",
    "(https://meltomeatlas.proteomics.wzw.tum.de/master_meltomeatlasapp/)\n",
    "* 1M variants \n",
    "* Label: temperature, meltpoint, fold_change, uniprot id \n",
    "\n",
    "**FireprotDB Stability** \n",
    "(https://loschmidt.chemi.muni.cz/fireprotdb/)\n",
    "* 53k variants\n",
    "* Label: ddG, dTm, pH, Tm, mutation_effect, uniprot id \n",
    "\n",
    "**ThermomutDB Stability**\n",
    "(https://biosig.lab.uq.edu.au/thermomutdb/downloads)\n",
    "* 12k variants\n",
    "* Label: pH, ddG, temperature, dTm, uniprot/pdb id \n",
    "\n",
    "**CAFA** \n",
    "(https://www.kaggle.com/competitions/cafa-5-protein-function-prediction/code)\n",
    "* 142k variants\n",
    "\n",
    "**Novozyme**\n",
    "(https://www.kaggle.com/code/jinyuansun/eda-and-finetune-esm)\n",
    "* 31k variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c9baf06",
   "metadata": {},
   "source": [
    "**Protsol Solubility**\n",
    "(https://huggingface.co/datasets/AI4Protein/ProtSolM)\n",
    "* 71k proteins\n",
    "* Label: solubility, no ID but has sequence\n",
    "* Unit: 0/1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "912e2d44",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PROTSOLM DATASET LOADING AND EXPORT \n",
    "from datasets import load_dataset\n",
    "ds = load_dataset(\"AI4Protein/ProtSolM\")\n",
    "# export each split\n",
    "ds[\"train\"].to_csv(\"protsolm_train.csv\")\n",
    "ds[\"test\"].to_csv(\"protsolm_test.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb14bb41",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (3135861269.py, line 12)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  Cell \u001b[1;32mIn[16], line 12\u001b[1;36m\u001b[0m\n\u001b[1;33m    meltome_path =\u001b[0m\n\u001b[1;37m                   ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#LOADING ALL BENCHMARK DATASETS AND MERGE IN ONE\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "\n",
    "PSI_PATH      = \"data/temp/sol_benchmark/PSI_Biology_solubility_trainset.csv\"\n",
    "NESG_PATH     = \"data/temp/sol_benchmark/nesg/nesg.csv\"\n",
    "PRICE_PATH    = \"data/temp/sol_benchmark/Price_usability_trainset.csv\"\n",
    "soluprot_train_path = \"data/temp/sol_benchmark/soluprot_data/training_set.csv\"\n",
    "soluprot_test_path = \"data/temp/sol_benchmark/soluprot_data/test_set.csv\" \n",
    "meltome_path = \"data/temp/stab_benchmark/meltome_cros-species.csv\"\n",
    "fireprot_path = \"data/temp/stab_benchmark/fireprotdb_results_stability.csv\"\n",
    "thermomutdb_path = \"data/temp/stab_benchmark/thermomutdb.json\"\n",
    "protsol_train_path = \"data/protsolm_data/protsolm_train.csv\"\n",
    "protsol_test_path = \"data/protsolm_data/protsolm_test.csv\"\n",
    "\n",
    "\n",
    "def read_fasta_dict(path: str):\n",
    "    seqs = {}\n",
    "    with open(path) as fh:\n",
    "        for header, seq in SimpleFastaParser(fh):\n",
    "            sid = header.split()[0].strip()\n",
    "            seqs[sid] = seq.strip()\n",
    "    return seqs\n",
    "\n",
    "def load_nesg(csv_path: str, fasta_path: str) -> pd.DataFrame:\n",
    "    df = pd.read_csv(csv_path)\n",
    "    seqs = read_fasta_dict(fasta_path)\n",
    "\n",
    "    df = df.rename(columns={\n",
    "        \"id\": \"nesg_id\",\n",
    "        \"exp\": \"nesg_exp\",\n",
    "        \"sol\": \"nesg_sol\",\n",
    "    })\n",
    "\n",
    "    df[\"sequence\"] = df[\"nesg_id\"].map(seqs)\n",
    "    df[\"dataset\"] = \"NESG\"\n",
    "    df[\"uniprot_id\"] = pd.NA\n",
    "\n",
    "    cols = [\"dataset\", \"uniprot_id\", \"sequence\", \"nesg_id\", \"nesg_exp\", \"nesg_sol\"]\n",
    "    return df[cols]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cda2ad4",
   "metadata": {},
   "source": [
    "# 3. The activity label of the PETase dataset \n",
    "\n",
    "**Docking** \n",
    "\n",
    "**PET catalysis rate** \n",
    "\n",
    "**PET-specific biophysical features** \n",
    "\n",
    "**Computational chemistry of PET and using it as a feature**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "364efbba",
   "metadata": {},
   "source": [
    "# 4. Fine-tuning esm2 model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "236b8815",
   "metadata": {},
   "source": [
    "# 5. Fine Tuning esm3 model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3bb20a9",
   "metadata": {},
   "source": [
    "# 6. Graph Neural Network"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
